lnwire/channel_announcement_2.go

```diff
@@ -0,0 +1,344 @@
+func (c *ChannelAnnouncement2) DigestToSign() (*chainhash.Hash, error) {
+	data, err := c.DataToSign()
+	if err != nil {
+		return nil, err
+	}
+
+	return MsgHash(chanAnn2MsgName, chanAnn2SigFieldName, data), nil
+}
+
+// DataToSign encodes the data to be signed into the ExtraOpaqueData member and
+// returns it.
+func (c *ChannelAnnouncement2) DataToSign() ([]byte, error) {
+	// The chain-hash record is only included if it is _not_ equal to the
+	// bitcoin mainnet genisis block hash.
+	var recordProducers []tlv.RecordProducer
+	if !c.ChainHash.Val.IsEqual(chaincfg.MainNetParams.GenesisHash) {
+		hash := tlv.ZeroRecordT[tlv.TlvType0, [32]byte]()
+		hash.Val = c.ChainHash.Val
+
+		recordProducers = append(recordProducers, &hash)
+	}
+
+	recordProducers = append(recordProducers,
+		&c.Features, &c.ShortChannelID, &c.Capacity, &c.NodeID1,
+		&c.NodeID2,
+	)
+
+	c.BitcoinKey1.WhenSome(func(key tlv.RecordT[tlv.TlvType12, [33]byte]) {
+		recordProducers = append(recordProducers, &key)
+	})
+
+	c.BitcoinKey2.WhenSome(func(key tlv.RecordT[tlv.TlvType14, [33]byte]) {
+		recordProducers = append(recordProducers, &key)
+	})
+
+	c.MerkleRootHash.WhenSome(
+		func(hash tlv.RecordT[tlv.TlvType16, [32]byte]) {
+			recordProducers = append(recordProducers, &hash)
+		},
+	)
+
+	err := EncodeMessageExtraData(&c.ExtraOpaqueData, recordProducers...)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.ExtraOpaqueData, nil
+}
+
+// MsgType returns the integer uniquely identifying this message type on the
+// wire.
+//
+// This is part of the lnwire.Message interface.
+func (c *ChannelAnnouncement2) MsgType() MessageType {
+	return MsgChannelAnnouncement2
+}
+
+// A compile time check to ensure ChannelAnnouncement2 implements the
+// lnwire.Message interface.
+var _ Message = (*ChannelAnnouncement2)(nil)
+
+// Node1KeyBytes returns the bytes representing the public key of node 1 in the
+// channel.
+//
+// NOTE: This is part of the ChannelAnnouncement interface.
+func (c *ChannelAnnouncement2) Node1KeyBytes() [33]byte {
+	return c.NodeID1.Val
+}
+
+// Node2KeyBytes returns the bytes representing the public key of node 2 in the
+// channel.
+//
+// NOTE: This is part of the ChannelAnnouncement interface.
+func (c *ChannelAnnouncement2) Node2KeyBytes() [33]byte {
+	return c.NodeID2.Val
+}
+
+// GetChainHash returns the hash of the chain which this channel's funding
+// transaction is confirmed in.
+//
+// NOTE: This is part of the ChannelAnnouncement interface.
+func (c *ChannelAnnouncement2) GetChainHash() chainhash.Hash {
+	return c.ChainHash.Val
+}
+
+// SCID returns the short channel ID of the channel.
+//
+// NOTE: This is part of the ChannelAnnouncement interface.
+func (c *ChannelAnnouncement2) SCID() ShortChannelID {
+	return c.ShortChannelID.Val
+}
+
+// Validate checks that the announcement signature is valid.
+//
+// NOTE: This is part of the ChannelAnnouncement interface.
+func (c *ChannelAnnouncement2) Validate(
```

Comment:
Very nice pattern! I do have two concerns here,
- the fact that we need to pass `fetchPKScript` here implies `lnwire` is not equipped to run validation logic. It may look pure, while in essence it's still an extension of `gossiper`(also not in favor of passing functions as params or configs as it's a hack to fix circular imports). This means if things change in the upper layer, the lower layer also needs to be aware, not good for modularization. We may be tempted to fix it by importing methods, say, from `lnwallet`, which leads to my second concert,
- now that we have validators built for gossip messages, what about others? Should we also need to validate channel open or update htlc messages in `lnwire`?

So I would prefer keeping `lnwire` as pure as it is now - it defines wire messages, handles encoding/decoding bytes, and leave the rest to be handled by the upper layers.

Reply:
updated 👍  lnwire now no longer contains the sign validation etc